---
title: "class11"
author: "Kelsey Fierro"
format: pdf
---

## Background

We saw lasat dau that pdb has 209,886 entries (Oct/Nov 2025).  UniProtKB (ie protein sequence database) has 199,579,901 entries

```{r}
209886/199579901 * 100
```

So the PDB has only 0.1% coverage of the main sequence database.

Enter AlphaFold database (AFDB)

Enter alphafold data base < https://alphafold.ebi.ac.uk > that attempts to provide computed models for all sequences in UniProt.

## AlphaFold

AlphaFold has 3 main outputs:

- the predicted coordinates (PDB files)
- A local quality score called **pLDDT** (one for each amino acid)
- A secondary quality score **PAE** Predicted Aligned Error (for each pair of amino acids)

We can run AlphaFold ourselves if we are not happy with AFDB (ie no coverage or poor model)

```{r}
results_dir <- "HIVPR_dimer_23119"
```
```{r}
# File names for all PDB models
pdb_files <- list.files(path=results_dir,
                        pattern="*.pdb",
                        full.names = TRUE)
```
```{r}
# Print PDB file names
basename(pdb_files)
```
```{r}
library(bio3d)
# Read all data from models and superpose/fit coords
pdbs <- pdbaln(pdb_files, fit=TRUE, exefile="msa")
```
```{r}
pdbs
```
```{r}
rd <- rmsd(pdbs, fit = T)
range(rd)
```
```{r}
library(pheatmap)
colnames(rd) <- paste0("m",1:5)
rownames(rd) <- paste0("m",1:5)
pheatmap(rd)
```
```{r}
# Read a reference PDB structure
pdb <- read.pdb("1hsg")
```
```{r}
plotb3(pdbs$b[1,], typ="l", lwd=2, sse=pdb)
points(pdbs$b[2,], typ="l", col="red")
points(pdbs$b[3,], typ="l", col="blue")
points(pdbs$b[4,], typ="l", col="darkgreen")
points(pdbs$b[5,], typ="l", col="orange")
abline(v=100, col="gray")
```
```{r}
core <- core.find(pdbs)
```
```{r}
core.inds <- print(core, vol=0.5)
```
```{r}
xyz <- pdbfit(pdbs, core.inds, outpath="corefit_structures")
```
# visualize the new "corefit_structures" in molstar

```{r}
rf <- rmsf(xyz)
plotb3(rf, sse=pdb)
abline(v=100, col="gray", ylab="RMSF")
```
```{r}
library(jsonlite)
# Listing of all PAE JSON files
pae_files <- list.files(path=results_dir,
                        pattern=".*model.*\\.json",
                        full.names = TRUE)
```
```{r}
pae1 <- read_json(pae_files[1],simplifyVector = TRUE)
pae5 <- read_json(pae_files[5],simplifyVector = TRUE)

attributes(pae1)
```
```{r}
# Per-residue pLDDT scores
# same as B-factor of PDB..
head(pae1$plddt)
```
```{r}
pae1$max_pae
```
```{r}
pae5$max_pae
```
```{r}
plot.dmat(pae1$pae,
          xlab="Residue Position (i)",
          ylab="Residue Position (j)")
```
```{r}
plot.dmat(pae5$pae, 
          xlab="Residue Position (i)",
          ylab="Residue Position (j)",
          grid.col = "black",
          zlim=c(0,30))
```
```{r}
plot.dmat(pae1$pae, 
          xlab="Residue Position (i)",
          ylab="Residue Position (j)",
          grid.col = "black",
          zlim=c(0,30))
```
```{r}
aln_file <- list.files(path=results_dir,
                       pattern=".a3m$",
                        full.names = TRUE)
aln_file
```
```{r}
aln <- read.fasta(aln_file[1], to.upper = TRUE)
```
```{r}
dim(aln$ali)
```
```{r}
sim <- conserv(aln)
plotb3(sim[1:99], sse=trim.pdb(pdb, chain="A"),
       ylab="Conservation Score")
```
```{r}
con <- consensus(aln, cutoff = 0.9)
con$seq
```
```{r}
m1.pdb <- read.pdb(pdb_files[1])
occ <- vec2resno(c(sim[1:99], sim[1:99]), m1.pdb$atom$resno)
write.pdb(m1.pdb, o=occ, file="m1_conserv.pdb")
```






