---
title: "class12"
author: "Kelsey Fierro"
format: pdf
---

Install DESeq2:
install.packages("BiocManager")
BiocManager::install("DESeq2")
```{r}
library(BiocManager)
```
```{r}
library(DESeq2)
```
## Background
Today we wull analyze some RNAseq data from Himes et al. on the effects of a common steroid 
(dexmethasone also called "dex") on airway smooth muscle cells (ASMs).
For this analysis we need two main inputs:
1. countData: a table of **counts** per gene (in rows) across experiments (in columns)
2. colData: **metadata** about the design of the experiments, the rows here must match the columns 
in `countData`

## Data Import
```{r}
counts <- read.csv("airway_scaledcounts.csv", row.names = 1)
metadata <- read.csv("airway_metadata.csv")
```

Let's have a wee peek at our `counts` data
```{r}
head(counts)
```
Add the `metadata`
```{r}
metadata
```
> Q1. How many "genes" are in this dataset?

```{r}
nrow(counts)
```

> Q2. How many experiments (i.e. columns in `counts` or rows in `metadata`) are there?

```{r}
ncol(counts)
```

> Q3. How many "control" experiments are there in the dataset?

```{r}
sum(metadata$dex == "control")
```

## Toy analysis example
1. Extract the "control" columns from `counts`
2. Calculate the mean value for each gene (rows) in these "control" columns
11/9/25, 11:25 PM 10/15/2025 - Posit Cloud
https://posit.cloud/spaces/707042/content/11139485 4/7
3-4. Do the same for the "treated" columns
5. Compare these mean values for each gene

Step 1.
```{r}
control.inds <- metadata$dex == "control"
control.counts <- counts[,control.inds]
head(control.counts)
```
Step 2.
```{r}
control.mean <- rowMeans(control.counts)
```

Step 3 and 4.
```{r}
treated.inds <- metadata$dex == "treated"
treated.counts <- counts[,treated.inds]
head(treated.counts)
```
```{r}
treated.mean <- rowMeans(treated.counts)
```

Step 5.
For ease of book-keeping we can store these together in one data frame called `meancounts`
```{r}
meancounts <- data.frame(control.mean, treated.mean)
head(meancounts)
```

Plot these against each other
```{r}
plot(meancounts)
```
This is screaming at me to log transform!
```{r}
plot(meancounts, log="xy")
```
We use log2 "fold-change" as a way to compare.
```{r}
#treated/control
log2(10/10)
log2(20/10)
log2(5/10)
log2(40/10)
```
```{r}
meancounts$log2fc <- log2(meancounts$treated.mean/meancounts$control.mean)
head(meancounts)
```
# Alternate method find zero values
```{r}
# 11/9/25, 11:25 PM 10/15/2025 - Posit Cloud
# https://posit.cloud/spaces/707042/content/11139485 5/7
x <- c(1,5,0,5)
which(x==0)
```
```{r}
y <- data.frame(a=c(1,5,0,5), b=c(1,0,5,5))
y
```
```{r}
which(y==0, arr.ind = T)
```
```{r}
zero.inds <- which(meancounts[,1:2]==0, arr.ind=T)[,1]
mygenes <- meancounts[-zero.inds,]
```

A common "rule-of-thumb" threshold for calling something "up" regulated is a log2-fold-change of 
+2 or greater. For down regulated -2 or less.

> Q. How many genes are "up" regulated at the +2 log2FC threshold?

```{r}
table(meancounts$log2fc>=2)
```
```{r}
sum(mygenes$log2fc>=2)
```

> Q. How many genes are "down" regulated at the -2 log2FC threshold?

```{r}
table(meancounts$log2fc<=2)
```

## DESeq analysis

Let's do this with DESeq2 and put some stats behind these numbers.

```{r, message=FALSE}
library(DESeq2)
```

DESeq wants 3 things for analysis, countData, colData, and design.
```{r}
dds <- DESeqDataSetFromMatrix(countData = counts,
                              colData = metadata,
                              design = ~dex)
```

The main function in the DESeq package to run analysis is called `DESeq()`.
```{r}
dds <- DESeq(dds)
```

Get the results out of this DESeq object with the function `results()`.
```{r}
res <- results(dds)
head(res)
```
# PCA

```{r}
vsd <- vst(dds, blind = FALSE)
plotPCA(vsd, intgroup = c("dex"))
```

```{r}
pcaData <- plotPCA(vsd, intgroup=c("dex"), returnData=TRUE)
head(pcaData)
```
```{r}
# Calculate percent variance per PC for the plot axis labels
percentVar <- round(100 * attr(pcaData, "percentVar"))
```

```{r}
library(ggplot2)
ggplot(pcaData) +
  aes(x = PC1, y = PC2, color = dex) +
  geom_point(size = 3) +
  xlab(paste0("PC1: ", percentVar[1], "% variance")) +
  ylab(paste0("PC2: ", percentVar[2], "% variance")) +
  coord_fixed() +
  theme_bw()
```


11/9/25, 11:25 PM 10/15/2025 - Posit Cloud
https://posit.cloud/spaces/707042/content/11139485 6/7

## Volcano Plot
This is a plot of log2FC vs adjusted p-value
```{r}
plot(res$log2FoldChange, -log(res$padj))
abline(v=c(-2,2), col="red")
abline(h=-log(0.05), col="red")
```

## Save our results
```{r}
write.csv(res, file="myresults.csv")
```

Make a nicer ggplot with color.
```{r}
library(ggplot2)
mycols <- rep("gray", nrow(res))
mycols[abs(res$log2FoldChange)>2] <- "blue"
mycols[res$padj>=0.05] <- "gray"
ggplot(res)+
  aes(log2FoldChange, -log(padj))+
  geom_point(col=mycols)
```
11/9/25, 11:25 PM 10/15/2025 - Posit Cloud

## Add anotation data

We need to add gene symbols, gene names, and other database ids to make my result useful for further analysis

```{r}
head(res)
```

We have ENSEMBLE database ids in our 'res' object

```{r}
head(rownames(res))
```

We can use the 'mapIds()' function from bioconductor to help us.

```{r}
library("AnnotationDbi")
library("org.Hs.eg.db")
```
Lets see what database id formats we can translate between

```{r}
columns(org.Hs.eg.db)
```

```{r}
res$symbol <- mapIds(org.Hs.eg.db,
                     keys=row.names(res), # our genenames
                     keytype="ENSEMBL", # the format of our genenames
                     column="SYMBOL") # the new format we want to add
head(res$symbol)
```

Add 'GENENAME' next
then 'ENTREZID'

```{r}
res$genename <- mapIds(org.Hs.eg.db,
                     keys=row.names(res), 
                     keytype="ENSEMBL", 
                     column="GENENAME") 
```
```{r}
res$entrezid <- mapIds(org.Hs.eg.db,
                     keys=row.names(res), 
                     keytype="ENSEMBL", 
                     column="ENTREZID") 
```
## Save my annotated results

```{r}
write.csv(res, file = "myresults_annotated.csv")
```

## Pathway analysis

We will use the **gage** function from bioconductor

```{r}
library(pathview)
library(gage)
library(gageData)

data(kegg.sets.hs)

# Examine the first 2 pathways in this kegg set for humans
head(kegg.sets.hs, 2)
```

What **gage** wants as input is a named vector of importance ie a vector with labeled fold changes.

```{r}
foldchanges = res$log2FoldChange
names(foldchanges) = res$entrez
head(foldchanges)
```
```{r}
# Get the results
keggres = gage(foldchanges, gsets=kegg.sets.hs)
```
```{r}
attributes(keggres)
```
```{r}
# Look at the top (less) pathways
head(keggres$less, 5)
```

Lets look at just one of these hsa05310
```{r, message=FALSE}
library(pathview)
```

```{r}
pathview(gene.data=foldchanges, pathway.id="hsa05310")
```
Insert figure for this pathway

![Asthma Pathway from KEGG with my differentially expressed genes highlighted](hsa05310.png)




